lib/core/config.rb

```diff
@@ -43,106 +70,176 @@ def should_app_start_with?(app_name)
+          "('allow_app_override_by_env' must be set to true in 'controlplane.yml'), " \
+          "or the --app command option."
   end
 
-  def ensure_current_config_org!(app_name)
+  def ensure_org!
     return if @org
 
-    raise "Can't find option 'cpln_org' for app '#{app_name}' in 'controlplane.yml', " \
-          "and CPLN_ORG env var is not set. " \
-          "The org can also be provided through --org."
+    raise "No org provided. " \
+          "The org can be provided either through the CPLN_ORG env var " \
+          "('allow_org_override_by_env' must be set to true in 'controlplane.yml'), " \
+          "the --org command option, " \
+          "or the 'cpln_org' key in 'controlplane.yml'."
   end
 
-  def ensure_config!
-    raise "'controlplane.yml' is empty." unless config
+  def ensure_required_options! # rubocop:disable Metrics/CyclomaticComplexity
+    ensure_app! if @required_options.include?(:app)
+    ensure_org! if @required_options.include?(:org) || app
+
+    missing_str = @required_options
+                  .reject { |option_name| %i[org app].include?(option_name) || options.key?(option_name) }
+                  .map { |option_name| "--#{option_name}" }
+                  .join(", ")
+
+    raise "Required options missing: #{missing_str}" unless missing_str.empty?
   end
 
-  def ensure_config_apps!
-    raise "Can't find key 'apps' in 'controlplane.yml'." unless config[:apps]
+  def new_config_keys
+    {
+      location: :default_location,
+      old_image_retention_days: :image_retention_days,
+      org: :cpln_org,
+      prefix: :match_if_app_name_starts_with
+    }
   end
 
-  def ensure_config_app!(app_name, app_options)
-    raise "App '#{app_name}' is empty in 'controlplane.yml'." unless app_options
+  def app_matches?(app_name1, app_name2, app_config)
+    app_name1 && app_name2 &&
+      (app_name1.to_s == app_name2.to_s ||
+        (app_config[:match_if_app_name_starts_with] && app_name1.to_s.start_with?(app_name2.to_s))
+      )
   end
 
-  def app_matches_current?(app_name, app_options)
-    app && (app_name.to_s == app || (app_options[:match_if_app_name_starts_with] && app.start_with?(app_name.to_s)))
+  def find_app_config(app_name1)
+    @app_configs ||= {}
+    @app_configs[app_name1] ||= apps.find do |app_name2, app_config|
+                                  app_matches?(app_name1, app_name2, app_config)
+                                end&.last
   end
 
-  def pick_current_config(app_name, app_options)
-    @current = app_options
-    ensure_current_config_app!(app_name)
+  def find_config_file
+    path = Pathname.new(".").expand_path
 
-    return if @org
+    loop do
+      config_file = path + CONFIG_FILE_LOCATION
+      break config_file if File.file?(config_file)
 
-    @org = current.fetch(:cpln_org)&.strip if current.key?(:cpln_org)
-    ensure_current_config_org!(app_name)
+      path = path.parent
+      if path.root?
```

Comment:
Please explain this look up logi and how it relates to the instructions.

Reply:
Let's handle this in a separate PR.